### PR TITLE
Add support for being used with sbot_simulator

### PR DESCRIPTION
**Pull request by WillB97**
- Adds support for the sbot_simulator method of attaching to boards over socket, instead of serial port.
- Adds support for remote sleep control to allow the the robot to sleep for periods of simulator time.
- Adds support for loading metadata from a file whose path is passed through an environment variable.
- Adds a WebotsRemoteCameraSource to receive camera frames over a socket.
- Adds ultrasound support to arduino.

Built for use with the SR fork: https://github.com/srobo/sbot_simulator

### CODE REVIEW

File: sr/robot3/arduino.py [hunk 1]
```diff
@@ -127,17 +135,46 @@ def _get_valid_board(
...
-                    f"Manually specified Arduino at port {serial_port!r}, "
-                    "could not be identified. Ignoring this device")
-            else:
-                logger.warning(
-                    f"Found Arduino-like serial port at {serial_port!r}, "
-                    "but it could not be identified. Ignoring this device")
+            if initial_identity is not None:
+                if initial_identity.board_type == 'manual':
+                    logger.warning(
+                        f"Manually specified Arduino at port {serial_port!r}, "
```
Comment (RealOrangeOne) [inline sr/robot3/arduino.py]:
```suggestion
                        f"Manually specified Arduino at port {serial_port!r} "
```

File: sr/robot3/arduino.py [hunk 2]
```diff
@@ -127,17 +135,46 @@ def _get_valid_board(
...
+                        "could not be identified. Ignoring this device")
+                elif initial_identity.manufacturer == 'sbot_simulator':
+                    logger.warning(
+                        f"Simulator specified arduino at port {serial_port!r}, "
```
Comment (RealOrangeOne) [inline sr/robot3/arduino.py]:
```suggestion
                        f"Simulator specified arduino at port {serial_port!r} "
```

File: sr/robot3/arduino.py [hunk 3]
```diff
@@ -127,17 +135,46 @@ def _get_valid_board(
...
+
+        :return: A mapping of board serial numbers to Arduinos
+        """
+        boards = {}
+        # The filter here is the name of the emulated board in the simulator
+        for board_info in get_simulator_boards('Arduino'):
+
+            # Create board identity from the info given
+            initial_identity = BoardIdentity(
+                manufacturer='sbot_simulator',
```
Comment (RealOrangeOne) [inline sr/robot3/arduino.py:168]: Nit: Should this be updated?

File: sr/robot3/robot.py [hunk 4]
```diff
@@ -350,7 +383,11 @@ def usbkey(self) -> Path:
 
         :returns: path to the mountpoint of the USB code drive.
         """
-        return self._astoria.fetch_mount_path()
+        if IN_SIMULATOR:
+            return Path(os.environ['SBOT_USBKEY_PATH'])
```
Comment (RealOrangeOne) [inline sr/robot3/robot.py:387]: Nit: Should this not be renamed?

File: sr/robot3/simulator/camera.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        """
+        The method for getting a new frame.
+
+        :param fresh: Whether to flush the device's buffer before capturing
+        the frame, unused.
+        """
+        self._serial.write(b"CAM:FRAME!\n")
+        # The image is encoded as a TLV (Type, Length, Value) packet
+        # so we need to read the header to get the type and length of the image
+        header = self._serial.read(HEADER_SIZE)
```
Comment (RealOrangeOne) [inline sr/robot3/simulator/camera.py:52]: Question: Should this use `_make_request`?

File: sr/robot3/arduino.py (hunk 3 above)
Comment (WillB97) [reply sr/robot3/arduino.py:168]: No, that's the name of the simulator.

File: sr/robot3/robot.py (hunk 4 above)
Comment (WillB97) [reply sr/robot3/robot.py:387]: No, it's the name of the variable provided by the simulator.

File: sr/robot3/simulator/camera.py (hunk 5 above)
Comment (WillB97) [reply sr/robot3/simulator/camera.py:52]: No, this is reading raw binary data.

**Review by 8BitJosh — APPROVED** [review]
Basic test in simulator and on a physical kit and all looks good

**Pull request merged**